File: arch/handle.go
package arch

import (
	"context"
	"log"
	"strings"

	"github.com/go-telegram-bot-api/telegram-bot-api"
)

const help = "*/arch* `package [repo]` , repo: eg: `stable` , `testing`, `aur` or `core` , `extra`"

type Arch struct{}

func (a *Arch) Handle(msg tgbotapi.Message, ctx context.Context, ch chan<- tgbotapi.Chattable) {
	args := strings.Split(msg.Text, " ")[1:]

	if len(args) == 0 {
		helpReply := tgbotapi.NewMessage(msg.Chat.ID, help)
		helpReply.ReplyToMessageID = msg.MessageID
		helpReply.ParseMode = tgbotapi.ModeMarkdown

		select {
		case <-ctx.Done():
		case ch <- helpReply:
		}
		return
	}

	var reply tgbotapi.MessageConfig

	if len(args) == 2 && strings.ToLower(args[1]) == "aur" {
		answer, err := aurQuery(args[0])

		switch err {
		default:
			log.Println(err)
			reply = tgbotapi.NewMessage(msg.Chat.ID, "哎呀，咱好像把这个 AUR 包吃了")

		case EmptyResult:
			reply = tgbotapi.NewMessage(msg.Chat.ID, "咱并没有找到这个 AUR 包，而且不是咱吃掉了！！！")

		case nil:
			str := answer.String()
			if str != "" {
				reply = tgbotapi.NewMessage(msg.Chat.ID, str)
			} else {
				log.Println("arch answer template execute failed")
				reply = tgbotapi.NewMessage(msg.Chat.ID, "哎呀，咱好像把这个 AUR 包吃了")
			}
		}
	} else {
		answer, err := officialQuery(args[0], args[1:]...)

		switch err {
		default:
			log.Println(err)
			reply = tgbotapi.NewMessage(msg.Chat.ID, "哎呀，咱好像把这个包吃了")

		case nil:
			str := answer.String()
			if str != "" {
				reply = tgbotapi.NewMessage(msg.Chat.ID, str)
			} else {
				log.Println("arch answer template execute failed")
				reply = tgbotapi.NewMessage(msg.Chat.ID, "哎呀，咱好像把这个包吃了")
			}

		case EmptyResult:
			if len(args[1:]) == 0 {
				answer, err := aurQuery(args[0])

				switch err {
				default:
					log.Println(err)
					reply = tgbotapi.NewMessage(msg.Chat.ID, "哎呀，咱好像把这个 AUR 包吃了")

				case EmptyResult:
					reply = tgbotapi.NewMessage(msg.Chat.ID, "咱并没有找到这个 AUR 包，而且不是咱吃掉了！！！")

				case nil:
					str := answer.String()
					if str != "" {
						reply = tgbotapi.NewMessage(msg.Chat.ID, str)
					} else {
						log.Println("arch answer template execute failed")
						reply = tgbotapi.NewMessage(msg.Chat.ID, "哎呀，咱好像把这个 AUR 包吃了")
					}
				}
			} else {
				reply = tgbotapi.NewMessage(msg.Chat.ID, "咱并没有找到这个包，而且不是咱吃掉了！！！")
			}
		}
	}

	reply.ReplyToMessageID = msg.MessageID
	reply.ParseMode = tgbotapi.ModeHTML

	select {
	case <-ctx.Done():
	case ch <- reply:
	}
}

File: src/handle.rs
use std::env;
use std::sync::LazyLock;

use teloxide::payloads::SendMessageSetters;
use teloxide::requests::{Requester, ResponseResult};
use teloxide::types::{Message, ParseMode};
use teloxide::Bot;

use crate::arch;
use crate::command::Command;
use crate::google;

static GOOGLE_KEY_AND_CX: LazyLock<Option<(String, String)>> = LazyLock::new(|| {
    match (
        env::var("HOLO_BOT_GOOGLE_KEY"),
        env::var("HOLO_BOT_GOOGLE_CX"),
    ) {
        (Ok(google_key), Ok(google_cx)) => Some((google_key, google_cx)),

        _ => None,
    }
});

struct Answer {
    result: String,
    mode: ParseMode,
}

pub async fn handle(bot: Bot, msg: Message, cmd: Command) -> ResponseResult<()> {
    let answer = match cmd {
        Command::Google(question) => {
            if let Some((google_key, google_cx)) = GOOGLE_KEY_AND_CX.as_ref() {
                if question.is_empty() {
                    Answer {
                        result: "*/google* `question`".to_string(),
                        mode: ParseMode::MarkdownV2,
                    }
                } else {
                    let answer = google::search(&question, google_key, google_cx)
                        .await
                        .map(|result| result.to_string())
                        .unwrap_or_else(|_| "哎呀咱好像在 Google 数据中心中迷路了~".to_string());

                    Answer {
                        result: answer,
                        mode: ParseMode::Html,
                    }
                }
            } else {
                return Ok(());
            }
        }

        Command::Arch(name, repos) => {
            match name {
                None => Answer {
                    result: "*/arch* `package [repo]`, repo: eg: `stable` , `testing`, `aur` or `core` , `extra`".to_string(),
                    mode: ParseMode::MarkdownV2,
                },
                Some(name) => {
                    let repos = match repos {
                        None => vec![],
                        Some(repos) => repos
                            .into_iter()
                            .filter(|repo| !repo.is_empty())
                            .collect::<Vec<String>>(),
                    };

                    let repos = repos.iter().map(|repo| repo.as_str()).collect::<Vec<_>>();

                    let answer = if repos.contains(&"aur") || repos.contains(&"AUR") {
                        arch::aur_query(&name)
                            .await
                            .map(|result| result.to_string())
                            .unwrap_or_else(|err| {
                                log::error!("{}", err);

                                "哎呀咱好像在 AUR 数据库中迷路了~".to_string()
                            })
                    } else if let Ok(result) = arch::official_query(&name, &repos).await {
                        if !result.is_empty() {
                            result.to_string()
                        } else {
                            arch::aur_query(&name)
                                .await
                                .map(|result| {
                                    if result.is_empty() {
                                        "咱没有找到结果，并且不是咱吃了！！！".to_string()
                                    } else {
                                        result.to_string()
                                    }
                                })
                                .unwrap_or_else(|err| {
                                    log::error!("{}", err);

                                    "哎呀咱好像在 AUR 数据库中迷路了~".to_string()
                                })
                        }
                    } else {
                        "哎呀咱好像在 Archlinux 数据库中迷路了~".to_string()
                    };

                    Answer {
                        result: answer,
                        mode: ParseMode::Html,
                    }
                }
            }
        }
    };

    bot.send_message(msg.chat.id, answer.result)
        .parse_mode(answer.mode)
        .reply_to_message_id(msg.id)
        .await?;

    Ok(())
}
